File: Sources/RecorderApp/MeetingIntelligence/MeetingIntelligencePipeline.swift
import Foundation

struct MeetingIntelligenceProgress: Equatable, Sendable {
    enum Stage: Equatable, Sendable {
        case summarizingChunks
        case reducingSummaries
        case generatingFinal
    }

    let stage: Stage
    let current: Int
    let total: Int
}

protocol MeetingIntelligenceGenerating: Sendable {
    func generate(
        transcript: TranscriptDocumentSnapshot,
        snapshot: OpenAICompatibleProviderSnapshot,
        onProgress: @escaping @Sendable (MeetingIntelligenceProgress) -> Void
    ) async throws -> MeetingIntelligenceGeneratedContent
}

enum MeetingIntelligencePipelineError: LocalizedError, Equatable, Sendable {
    case sourceTooLarge
    case invalidSource
    case tooManyChunks
    case partialTooLarge
    case tooManyRequests
    case maximumDepthReached
    case deadlineExceeded
    case cancelled

    var errorDescription: String? {
        switch self {
        case .sourceTooLarge: "The transcript is too large to summarize."
        case .invalidSource: "The transcript is empty or invalid."
        case .tooManyChunks: "The transcript requires too many summary chunks."
        case .partialTooLarge: "The provider returned an oversized partial summary."
        case .tooManyRequests: "The summary request limit was reached."
        case .maximumDepthReached: "The summary reduction limit was reached."
        case .deadlineExceeded: "The summary took too long to complete."
        case .cancelled: "Meeting intelligence was cancelled."
        }
    }
}

struct MeetingIntelligencePipeline: MeetingIntelligenceGenerating {
    typealias Now = @Sendable () -> ContinuousClock.Instant

    private enum Limits {
        static let sourceBytes = 4 * 1_024 * 1_024
        static let chunkBytes = 64 * 1_024
        static let maximumChunks = 64
        static let partialBytes = 4 * 1_024
        static let reductionFanIn = 12
        static let maximumDepth = 4
        static let maximumRequests = 71
        static let maximumDuration = Duration.seconds(1_800)
    }

    private let client: any MeetingIntelligenceRequesting
    private let now: Now

    init(
        client: any MeetingIntelligenceRequesting,
        now: @escaping Now = { ContinuousClock().now }
    ) {
        self.client = client
        self.now = now
    }

    func generate(
        transcript: TranscriptDocumentSnapshot,
        snapshot: OpenAICompatibleProviderSnapshot,
        onProgress: @escaping @Sendable (MeetingIntelligenceProgress) -> Void
    ) async throws -> MeetingIntelligenceGeneratedContent {
        let deadline = now().advanced(by: Limits.maximumDuration)
        try checkBoundary(deadline: deadline)
        guard transcript.data.count <= Limits.sourceBytes else {
            throw MeetingIntelligencePipelineError.sourceTooLarge
        }
        guard !transcript.data.isEmpty,
              let source = String(data: transcript.data, encoding: .utf8) else {
            throw MeetingIntelligencePipelineError.invalidSource
        }

        let chunks = try split(source)
        guard chunks.count <= Limits.maximumChunks else {
            throw MeetingIntelligencePipelineError.tooManyChunks
        }

        var requestCount = 0
        if chunks.count == 1 {
            try checkBoundary(deadline: deadline)
            try reserveRequest(&requestCount)
            onProgress(.init(stage: .generatingFinal, current: 1, total: 1))
            let final = try await client.requestFinalResult(input: chunks[0], snapshot: snapshot)
            try checkBoundary(deadline: deadline)
            return final
        }

        var summaries: [String] = []
        summaries.reserveCapacity(chunks.count)
        for (index, chunk) in chunks.enumerated() {
            try checkBoundary(deadline: deadline)
            try reserveRequest(&requestCount)
            onProgress(.init(
                stage: .summarizingChunks,
                current: index + 1,
                total: chunks.count
            ))
            let summary = try await client.requestPartialSummary(input: chunk, snapshot: snapshot)
            try checkBoundary(deadline: deadline)
            try validatePartial(summary)
            summaries.append(summary)
        }

        var depth = 0
        while summaries.count > Limits.reductionFanIn || joinedByteCount(summaries) > Limits.chunkBytes {
            guard depth < Limits.maximumDepth else {
                throw MeetingIntelligencePipelineError.maximumDepthReached
            }
            let groups = stride(from: 0, to: summaries.count, by: Limits.reductionFanIn)
                .map { Array(summaries[$0 ..< min($0 + Limits.reductionFanIn, summaries.count)]) }
            var reduced: [String] = []
            reduced.reserveCapacity(groups.count)
            for (index, group) in groups.enumerated() {
                let input = group.joined(separator: "\n")
                try checkBoundary(deadline: deadline)
                try reserveRequest(&requestCount)
                onProgress(.init(
                    stage: .reducingSummaries,
                    current: index + 1,
                    total: groups.count
                ))
                let summary = try await client.requestPartialSummary(input: input, snapshot: snapshot)
                try checkBoundary(deadline: deadline)
                try validatePartial(summary)
                reduced.append(summary)
            }
            summaries = reduced
            depth += 1
        }

        let finalInput = summaries.joined(separator: "\n")
        try checkBoundary(deadline: deadline)
        try reserveRequest(&requestCount)
        onProgress(.init(stage: .generatingFinal, current: 1, total: 1))
        let final = try await client.requestFinalResult(input: finalInput, snapshot: snapshot)
        try checkBoundary(deadline: deadline)
        return final
    }

    private func reserveRequest(_ count: inout Int) throws {
        guard count < Limits.maximumRequests else {
            throw MeetingIntelligencePipelineError.tooManyRequests
        }
        count += 1
    }

    private func checkBoundary(deadline: ContinuousClock.Instant) throws {
        if Task.isCancelled {
            throw MeetingIntelligencePipelineError.cancelled
        }
        guard now() < deadline else {
            throw MeetingIntelligencePipelineError.deadlineExceeded
        }
    }

    private func validatePartial(_ summary: String) throws {
        guard !summary.isEmpty, summary.utf8.count <= Limits.partialBytes else {
            throw MeetingIntelligencePipelineError.partialTooLarge
        }
    }

    private func joinedByteCount(_ values: [String]) -> Int {
        values.reduce(max(0, values.count - 1)) { $0 + $1.utf8.count }
    }

    private func split(_ source: String) throws -> [String] {
        let data = Data(source.utf8)
        var offset = 0
        var chunks: [String] = []
        while offset < data.count {
            let upper = min(offset + Limits.chunkBytes, data.count)
            let end = preferredBoundary(in: data, start: offset, upper: upper)
            guard end > offset,
                  let chunk = String(data: data[offset ..< end], encoding: .utf8) else {
                throw MeetingIntelligencePipelineError.invalidSource
            }
            chunks.append(chunk)
            offset = end
        }
        return chunks
    }

    private func preferredBoundary(in data: Data, start: Int, upper: Int) -> Int {
        guard upper < data.count else { return upper }
        let scalarEnd = scalarBoundary(in: data, start: start, upper: upper)
        let bytes = data[start ..< scalarEnd]
        for separator in [Data("\n\n".utf8), Data("\n".utf8), Data(". ".utf8), Data("! ".utf8), Data("? ".utf8)] {
            if let range = bytes.range(of: separator, options: .backwards) {
                return start + range.upperBound
            }
        }
        return scalarEnd
    }

    private func scalarBoundary(in data: Data, start: Int, upper: Int) -> Int {
        var index = upper
        while index > start && (data[index] & 0b1100_0000) == 0b1000_0000 {
            index -= 1
        }
        return index == start ? upper : index
    }
}

File: windows/src/Recorder.Application/Transcription/MeetingIntelligencePipeline.cs
using System.Text;
using TeamsRecorder.Windows.Application.AI;

namespace TeamsRecorder.Windows.Application.Transcription;

public sealed record MeetingIntelligenceTranscriptRevision(string Sha256, int ByteCount)
{
    public static MeetingIntelligenceTranscriptRevision FromBytes(ReadOnlySpan<byte> bytes)
    {
        var digest = Convert.ToHexStringLower(System.Security.Cryptography.SHA256.HashData(bytes));
        return new("sha256:" + digest, bytes.Length);
    }
}

public sealed record MeetingIntelligenceCanonicalTranscript(
    string Text,
    byte[] Utf8Bytes,
    MeetingIntelligenceTranscriptRevision Revision);

public enum MeetingIntelligencePipelineStage { SummarizingChunks, ReducingSummaries, GeneratingFinal }
public sealed record MeetingIntelligenceProgress(MeetingIntelligencePipelineStage Stage, int Current, int Total);

public enum MeetingIntelligencePipelineFailure
{
    SourceTooLarge,
    InvalidSource,
    RequestTooLarge,
    TooManyChunks,
    PartialTooLarge,
    TooManyRequests,
    MaximumDepthReached,
    DeadlineExceeded,
}

public sealed class MeetingIntelligencePipelineException(MeetingIntelligencePipelineFailure failure, string message)
    : Exception(message)
{
    public MeetingIntelligencePipelineFailure Failure { get; } = failure;
}

/// <summary>
/// Deterministic, bounded long-transcript map/reduce pipeline. It never truncates source or
/// publishes partial work. A worst-case 64-chunk job requires at most 64 + 6 + 1 requests.
/// </summary>
public sealed class MeetingIntelligencePipeline
{
    public const int MaximumSourceBytes = 4 * 1024 * 1024;
    public const int MaximumChunkBytes = 64 * 1024;
    public const int MaximumChunks = 64;
    public const int MaximumPartialSummaryBytes = 4 * 1024;
    public const int MaximumReductionGroupSize = 12;
    public const int MaximumReductionDepth = 4;
    public const int MaximumRequests = 71;
    public static readonly TimeSpan MaximumDuration = TimeSpan.FromMinutes(30);

    private readonly IMeetingIntelligenceRequestClient client;
    private readonly Func<DateTimeOffset> now;

    public MeetingIntelligencePipeline(
        IMeetingIntelligenceRequestClient client,
        Func<DateTimeOffset>? now = null)
    {
        this.client = client ?? throw new ArgumentNullException(nameof(client));
        this.now = now ?? (() => DateTimeOffset.UtcNow);
    }

    public async Task<MeetingIntelligenceGeneratedContent> GenerateAsync(
        MeetingIntelligenceCanonicalTranscript transcript,
        OpenAICompatibleProviderSnapshot snapshot,
        IProgress<MeetingIntelligenceProgress>? progress = null,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(transcript);
        ArgumentNullException.ThrowIfNull(snapshot);
        var deadline = now() + MaximumDuration;
        CheckBoundary(deadline, cancellationToken);
        if (transcript.Utf8Bytes.Length > MaximumSourceBytes)
            throw new MeetingIntelligencePipelineException(MeetingIntelligencePipelineFailure.SourceTooLarge, "The canonical transcript exceeds 4 MiB.");
        if (transcript.Utf8Bytes.Length == 0 || !string.Equals(transcript.Text, StrictUtf8(transcript.Utf8Bytes), StringComparison.Ordinal))
            throw new MeetingIntelligencePipelineException(MeetingIntelligencePipelineFailure.InvalidSource, "The canonical transcript is empty or invalid UTF-8.");

        var chunks = Split(transcript.Utf8Bytes, snapshot);
        if (chunks.Count > MaximumChunks)
            throw new MeetingIntelligencePipelineException(MeetingIntelligencePipelineFailure.TooManyChunks, "The transcript requires more than 64 chunks.");

        var requestCount = 0;
        if (chunks.Count == 1)
        {
            PrepareRequest(ref requestCount, deadline, cancellationToken, progress,
                new(MeetingIntelligencePipelineStage.GeneratingFinal, 1, 1));
            var direct = await client.RequestFinalResultAsync(chunks[0], snapshot, cancellationToken).ConfigureAwait(false);
            CheckBoundary(deadline, cancellationToken);
            return ValidateFinal(direct);
        }

        var summaries = new List<string>(chunks.Count);
        for (var index = 0; index < chunks.Count; index++)
        {
            PrepareRequest(ref requestCount, deadline, cancellationToken, progress,
                new(MeetingIntelligencePipelineStage.SummarizingChunks, index + 1, chunks.Count));
            var summary = await client.RequestPartialSummaryAsync(chunks[index], snapshot, cancellationToken).ConfigureAwait(false);
            CheckBoundary(deadline, cancellationToken);
            summaries.Add(ValidatePartial(summary));
        }

        var depth = 0;
        while (RequiresReduction(summaries, snapshot))
        {
            if (depth >= MaximumReductionDepth)
                throw new MeetingIntelligencePipelineException(MeetingIntelligencePipelineFailure.MaximumDepthReached, "The summary reduction exceeded four levels.");
            var groups = ReductionGroups(summaries, snapshot);
            var reduced = new List<string>(groups.Count);
            for (var index = 0; index < groups.Count; index++)
            {
                PrepareRequest(ref requestCount, deadline, cancellationToken, progress,
                    new(MeetingIntelligencePipelineStage.ReducingSummaries, index + 1, groups.Count));
                var summary = await client.RequestPartialSummaryAsync(string.Join('\n', groups[index]), snapshot, cancellationToken).ConfigureAwait(false);
                CheckBoundary(deadline, cancellationToken);
                reduced.Add(ValidatePartial(summary));
            }
            summaries = reduced;
            depth++;
        }

        var finalInput = string.Join('\n', summaries);
        if (!client.FitsRequest(finalInput, snapshot, final: true))
            throw new MeetingIntelligencePipelineException(MeetingIntelligencePipelineFailure.RequestTooLarge, "The reduced summary cannot fit in the final request.");
        PrepareRequest(ref requestCount, deadline, cancellationToken, progress,
            new(MeetingIntelligencePipelineStage.GeneratingFinal, 1, 1));
        var result = await client.RequestFinalResultAsync(finalInput, snapshot, cancellationToken).ConfigureAwait(false);
        CheckBoundary(deadline, cancellationToken);
        return ValidateFinal(result);
    }

    private IReadOnlyList<string> Split(byte[] bytes, OpenAICompatibleProviderSnapshot snapshot)
    {
        var chunks = new List<string>();
        var offset = 0;
        var isSingleChunk = bytes.Length <= MaximumChunkBytes;
        while (offset < bytes.Length)
        {
            if (chunks.Count >= MaximumChunks)
                throw new MeetingIntelligencePipelineException(MeetingIntelligencePipelineFailure.TooManyChunks, "The transcript requires more than 64 chunks.");
            var upper = Math.Min(offset + MaximumChunkBytes, bytes.Length);
            var remainingSlots = MaximumChunks - chunks.Count - 1;
            var lower = Math.Max(offset + 1, bytes.Length - (remainingSlots * MaximumChunkBytes));
            var preferred = PreferredBoundary(bytes, offset, Math.Min(lower, upper), upper);
            var end = FittingBoundary(bytes, offset, Math.Min(lower, upper), preferred, snapshot, final: isSingleChunk);
            if (end <= offset)
                throw new MeetingIntelligencePipelineException(MeetingIntelligencePipelineFailure.RequestTooLarge, "A transcript chunk cannot fit in the request envelope.");
            var value = StrictUtf8(bytes.AsSpan(offset, end - offset));
            if (value is null)
                throw new MeetingIntelligencePipelineException(MeetingIntelligencePipelineFailure.InvalidSource, "The transcript contains invalid UTF-8.");
            chunks.Add(value);
            offset = end;
        }
        return chunks;
    }

    private int FittingBoundary(byte[] bytes, int start, int lower, int upper, OpenAICompatibleProviderSnapshot snapshot, bool final)
    {
        var boundaries = new List<int> { start };
        for (var index = start + 1; index <= upper; index++)
        {
            if (index == bytes.Length || !IsContinuation(bytes[index])) boundaries.Add(index);
        }
        var low = boundaries.FindIndex(value => value >= lower);
        if (low < 1) low = 1;
        var high = boundaries.Count - 1;
        var best = -1;
        while (low <= high)
        {
            var middle = low + ((high - low) / 2);
            var candidate = boundaries[middle];
            var value = StrictUtf8(bytes.AsSpan(start, candidate - start));
            if (value is not null && client.FitsRequest(value, snapshot, final))
            {
                best = candidate;
                low = middle + 1;
            }
            else high = middle - 1;
        }
        if (best < 0)
            throw new MeetingIntelligencePipelineException(MeetingIntelligencePipelineFailure.RequestTooLarge, "A transcript scalar cannot fit in the request envelope.");
        return best;
    }

    private static int PreferredBoundary(byte[] bytes, int start, int lower, int upper)
    {
        if (upper == bytes.Length) return upper;
        while (upper > start && IsContinuation(bytes[upper])) upper--;
        var preferredLower = Math.Max(lower, Math.Max(start + 1, upper - (16 * 1024)));
        for (var index = upper - 1; index >= preferredLower; index--)
        {
            if (bytes[index] == (byte)'\n') return index + 1;
            if (index + 1 < upper && bytes[index + 1] == (byte)' ' && bytes[index] is (byte)'.' or (byte)'!' or (byte)'?') return index + 2;
        }
        return upper;
    }

    private bool RequiresReduction(IReadOnlyList<string> summaries, OpenAICompatibleProviderSnapshot snapshot)
    {
        if (summaries.Count > MaximumReductionGroupSize) return true;
        var joined = string.Join('\n', summaries);
        return Encoding.UTF8.GetByteCount(joined) > MaximumChunkBytes || !client.FitsRequest(joined, snapshot, final: true);
    }

    private IReadOnlyList<IReadOnlyList<string>> ReductionGroups(IReadOnlyList<string> summaries, OpenAICompatibleProviderSnapshot snapshot)
    {
        var groups = new List<IReadOnlyList<string>>();
        var current = new List<string>();
        foreach (var summary in summaries)
        {
            var candidate = current.Append(summary).ToArray();
            if (candidate.Length <= MaximumReductionGroupSize &&
                Encoding.UTF8.GetByteCount(string.Join('\n', candidate)) <= MaximumChunkBytes &&
                client.FitsRequest(string.Join('\n', candidate), snapshot, final: false))
            {
                current.Add(summary);
                continue;
            }
            if (current.Count == 0 || !client.FitsRequest(summary, snapshot, final: false))
                throw new MeetingIntelligencePipelineException(MeetingIntelligencePipelineFailure.RequestTooLarge, "A partial summary cannot fit in a reduction request.");
            groups.Add(current.ToArray());
            current.Clear();
            current.Add(summary);
        }
        if (current.Count > 0) groups.Add(current.ToArray());
        return groups;
    }

    private static string ValidatePartial(string value)
    {
        try { return MeetingIntelligenceOutputValidator.ValidateSummary(value, MaximumPartialSummaryBytes); }
        catch (MeetingIntelligenceClientException error)
        {
            throw new MeetingIntelligencePipelineException(MeetingIntelligencePipelineFailure.PartialTooLarge, error.Message);
        }
    }

    private static MeetingIntelligenceGeneratedContent ValidateFinal(MeetingIntelligenceGeneratedContent value) => new(
        MeetingIntelligenceOutputValidator.ValidateTitle(value.SuggestedTitle),
        MeetingIntelligenceOutputValidator.ValidateSummary(value.Summary));

    private void PrepareRequest(
        ref int requestCount,
        DateTimeOffset deadline,
        CancellationToken cancellationToken,
        IProgress<MeetingIntelligenceProgress>? progress,
        MeetingIntelligenceProgress value)
    {
        CheckBoundary(deadline, cancellationToken);
        if (requestCount >= MaximumRequests)
            throw new MeetingIntelligencePipelineException(MeetingIntelligencePipelineFailure.TooManyRequests, "The job exceeded 71 provider requests.");
        requestCount++;
        progress?.Report(value);
        CheckBoundary(deadline, cancellationToken);
    }

    private void CheckBoundary(DateTimeOffset deadline, CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        if (now() >= deadline)
            throw new MeetingIntelligencePipelineException(MeetingIntelligencePipelineFailure.DeadlineExceeded, "Meeting intelligence exceeded 30 minutes.");
    }

    private static bool IsContinuation(byte value) => (value & 0b1100_0000) == 0b1000_0000;

    private static string? StrictUtf8(ReadOnlySpan<byte> bytes)
    {
        try { return new UTF8Encoding(false, true).GetString(bytes); }
        catch (DecoderFallbackException) { return null; }
    }
}
